File: macos/Sources/MacDroidSyncCore/CryptoBox.swift
import CryptoKit
import Foundation

/// Key derivation and AES-256-GCM sealing shared with the Android client.
///
/// key   = HKDF-SHA256(ikm: normalized pairing code, salt: "MacDroidSync/v1",
///                     info: "clipboard-channel", length: 32)
/// nonce = 4 random session bytes || 8 bytes big-endian frame counter
/// body  = nonce || ciphertext || 16 byte GCM tag
public enum CryptoBox {
    public static let hkdfSalt = Data("MacDroidSync/v1".utf8)
    public static let hkdfInfo = Data("clipboard-channel".utf8)
    public static let nonceLength = 12
    public static let tagLength = 16

    /// Pairing codes are compared case insensitively and free of separators,
    /// so "abcd-efgh" and "ABCDEFGH" derive the same key.
    public static func normalize(pairingCode: String) -> String {
        String(pairingCode.uppercased().filter { $0.isLetter || $0.isNumber })
    }

    public static func deriveKey(pairingCode: String) -> SymmetricKey {
        deriveKey(pairingCode: pairingCode, info: hkdfInfo, outputByteCount: 32)
    }

    /// Same pairing code, different purpose: `info` separates the channel key
    /// from the presence beacon material (see `PresenceBeacon`).
    public static func deriveKey(pairingCode: String, info: Data, outputByteCount: Int) -> SymmetricKey {
        let material = SymmetricKey(data: Data(normalize(pairingCode: pairingCode).utf8))
        return HKDF<SHA256>.deriveKey(
            inputKeyMaterial: material,
            salt: hkdfSalt,
            info: info,
            outputByteCount: outputByteCount
        )
    }

    /// Deterministic sealing, used by the cross platform test vectors.
    public static func seal(plaintext: Data, key: SymmetricKey, nonce: Data) throws -> Data {
        let box = try AES.GCM.seal(plaintext, using: key, nonce: AES.GCM.Nonce(data: nonce))
        var out = Data(nonce)
        out.append(box.ciphertext)
        out.append(box.tag)
        return out
    }

    public static func open(body: Data, key: SymmetricKey) throws -> Data {
        guard body.count > nonceLength + tagLength else {
            throw ProtocolError.malformedFrame("sealed body of \(body.count) bytes is too short")
        }
        let bytes = Data(body)
        let nonce = bytes.prefix(nonceLength)
        let sealed = bytes.dropFirst(nonceLength)
        let ciphertext = sealed.dropLast(tagLength)
        let tag = sealed.suffix(tagLength)
        do {
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: nonce),
                ciphertext: Data(ciphertext),
                tag: Data(tag)
            )
            return try AES.GCM.open(box, using: key)
        } catch {
            throw ProtocolError.authenticationFailed
        }
    }

    /// Human readable pairing code, 16 characters of a confusable free
    /// alphabet grouped as XXXX-XXXX-XXXX-XXXX (80 bits of entropy).
    public static func generatePairingCode() -> String {
        let alphabet = Array("ABCDEFGHJKLMNPQRSTUVWXYZ23456789")
        var characters: [Character] = []
        var random = SystemRandomNumberGenerator()
        for index in 0 ..< 16 {
            if index > 0 && index % 4 == 0 { characters.append("-") }
            characters.append(alphabet[Int(random.next(upperBound: UInt64(alphabet.count)))])
        }
        return String(characters)
    }

    public static func fingerprint(of text: String) -> String {
        let digest = SHA256.hash(data: Data(text.utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}

/// Per connection sealing state: an outgoing nonce counter plus replay
/// protection for the sequence numbers seen on the wire.
public final class FrameCodec {
    private let key: SymmetricKey
    private let noncePrefix: Data
    private var sendCounter: UInt64 = 0
    private var lastSeenSeq: UInt64 = 0
    private var outgoingSeq: UInt64 = 0

    public init(pairingCode: String) {
        self.key = CryptoBox.deriveKey(pairingCode: pairingCode)
        var prefix = Data(count: 4)
        prefix.withUnsafeMutableBytes { buffer in
            _ = SecRandomCopyBytes(kSecRandomDefault, 4, buffer.baseAddress!)
        }
        self.noncePrefix = prefix
    }

    public func nextSequence() -> UInt64 {
        outgoingSeq += 1
        return outgoingSeq
    }

    public func seal(_ plaintext: Data) throws -> Data {
        sendCounter += 1
        var nonce = Data(noncePrefix)
        var counter = sendCounter.bigEndian
        withUnsafeBytes(of: &counter) { nonce.append(contentsOf: $0) }
        return try CryptoBox.seal(plaintext: plaintext, key: key, nonce: nonce)
    }

    /// Decrypts a frame body and rejects replayed or reordered messages.
    public func open(_ body: Data) throws -> Message {
        let plaintext = try CryptoBox.open(body: body, key: key)
        let message = try Message.decode(plaintext)
        guard message.seq > lastSeenSeq else {
            throw ProtocolError.replayDetected(message.seq)
        }
        lastSeenSeq = message.seq
        return message
    }
}

File: android/app/src/main/java/pl/wojas/macdroidsync/CryptoBox.kt
package pl.wojas.macdroidsync

import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Key derivation and AES-256-GCM sealing, byte compatible with the CryptoKit
 * based macOS implementation. See PROTOCOL.md for the test vectors.
 */
object CryptoBox {
    private val HKDF_SALT = "MacDroidSync/v1".toByteArray(Charsets.UTF_8)
    private val HKDF_INFO = "clipboard-channel".toByteArray(Charsets.UTF_8)
    const val NONCE_LENGTH = 12
    const val TAG_LENGTH = 16

    /** Uppercase, separators removed: "abcd-efgh" and "ABCDEFGH" are the same code. */
    fun normalize(pairingCode: String): String =
        pairingCode.uppercase().filter { it.isLetterOrDigit() }

    fun deriveKey(pairingCode: String): ByteArray =
        hkdfSha256(
            ikm = normalize(pairingCode).toByteArray(Charsets.UTF_8),
            salt = HKDF_SALT,
            info = HKDF_INFO,
            length = 32,
        )

    fun seal(plaintext: ByteArray, key: ByteArray, nonce: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_LENGTH * 8, nonce))
        val sealed = cipher.doFinal(plaintext)
        return nonce + sealed
    }

    fun open(body: ByteArray, key: ByteArray): ByteArray {
        if (body.size <= NONCE_LENGTH + TAG_LENGTH) {
            throw ProtocolException("Sealed body of ${body.size} bytes is too short")
        }
        val nonce = body.copyOfRange(0, NONCE_LENGTH)
        val sealed = body.copyOfRange(NONCE_LENGTH, body.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_LENGTH * 8, nonce))
        return try {
            cipher.doFinal(sealed)
        } catch (error: Exception) {
            throw ProtocolException("Authentication failed - the pairing code does not match")
        }
    }

    fun fingerprint(text: String): String {
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        return digest.digest(text.toByteArray(Charsets.UTF_8)).joinToString("") { "%02x".format(it) }
    }

    fun randomBytes(count: Int): ByteArray = ByteArray(count).also { SecureRandom().nextBytes(it) }

    private fun hkdfSha256(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        val extract = Mac.getInstance("HmacSHA256")
        extract.init(SecretKeySpec(salt, "HmacSHA256"))
        val prk = extract.doFinal(ikm)

        val expand = Mac.getInstance("HmacSHA256")
        expand.init(SecretKeySpec(prk, "HmacSHA256"))
        val output = ByteArray(length)
        var block = ByteArray(0)
        var offset = 0
        var counter = 1
        while (offset < length) {
            expand.reset()
            expand.update(block)
            expand.update(info)
            expand.update(counter.toByte())
            block = expand.doFinal()
            val take = minOf(block.size, length - offset)
            block.copyInto(output, offset, 0, take)
            offset += take
            counter++
        }
        return output
    }
}

/**
 * Per connection sealing state: an outgoing nonce counter plus replay protection
 * for the sequence numbers seen on the wire.
 */
class FrameCodec(pairingCode: String) {
    private val key = CryptoBox.deriveKey(pairingCode)
    private val noncePrefix = CryptoBox.randomBytes(4)
    private var sendCounter = 0L
    private var lastSeenSeq = 0L
    private var outgoingSeq = 0L

    fun nextSequence(): Long = ++outgoingSeq

    fun seal(message: Message): ByteArray {
        sendCounter++
        val nonce = ByteArray(CryptoBox.NONCE_LENGTH)
        noncePrefix.copyInto(nonce)
        for (index in 0 until 8) {
            nonce[4 + index] = (sendCounter ushr (8 * (7 - index)) and 0xFF).toByte()
        }
        return CryptoBox.seal(message.toBytes(), key, nonce)
    }

    /** Decrypts a frame body and rejects replayed or reordered messages. */
    fun open(body: ByteArray): Message {
        val message = Message.parse(CryptoBox.open(body, key))
        if (message.seq <= lastSeenSeq) {
            throw ProtocolException("Replayed or out of order message (seq ${message.seq})")
        }
        lastSeenSeq = message.seq
        return message
    }
}
